File: checkers.go
// checkers.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const EMPTY = 0
const WHITE = 1
const BLACK = 2

type Checkers struct {
	board [8][8]int
	turn  int
}

func NewCheckers() *Checkers {
	c := &Checkers{turn: WHITE}
	// расстановка
	for row := 0; row < 8; row++ {
		for col := 0; col < 8; col++ {
			if (row+col)%2 == 1 {
				if row < 3 {
					c.board[row][col] = WHITE
				} else if row > 4 {
					c.board[row][col] = BLACK
				}
			}
		}
	}
	return c
}

func (c *Checkers) PrintBoard() {
	fmt.Println("  a b c d e f g h")
	for row := 0; row < 8; row++ {
		fmt.Printf("%d ", row+1)
		for col := 0; col < 8; col++ {
			ch := '·'
			if c.board[row][col] == WHITE {
				ch = '○'
			} else if c.board[row][col] == BLACK {
				ch = '●'
			}
			fmt.Printf("%c ", ch)
		}
		fmt.Println()
	}
}

func (c *Checkers) isValidPos(row, col int) bool {
	return row >= 0 && row < 8 && col >= 0 && col < 8
}

func (c *Checkers) getMoves(row, col int) [][3][2]int {
	piece := c.board[row][col]
	if piece == EMPTY {
		return nil
	}
	dir := 1
	if piece == BLACK {
		dir = -1
	}
	var moves [][3][2]int
	// простые ходы
	for _, dc := range []int{-1, 1} {
		nr, nc := row+dir, col+dc
		if c.isValidPos(nr, nc) && c.board[nr][nc] == EMPTY {
			moves = append(moves, [3][2]int{{row, col}, {nr, nc}, {-1, -1}})
		}
	}
	// захваты
	for _, dc := range []int{-1, 1} {
		nr, nc := row+dir*2, col+dc*2
		if c.isValidPos(nr, nc) {
			mr, mc := row+dir, col+dc
			if c.isValidPos(mr, mc) && c.board[mr][mc] != EMPTY && c.board[mr][mc] != piece && c.board[nr][nc] == EMPTY {
				moves = append(moves, [3][2]int{{row, col}, {nr, nc}, {mr, mc}})
			}
		}
	}
	return moves
}

func (c *Checkers) getAllMoves(color int) [][3][2]int {
	var moves [][3][2]int
	for row := 0; row < 8; row++ {
		for col := 0; col < 8; col++ {
			if c.board[row][col] == color {
				moves = append(moves, c.getMoves(row, col)...)
			}
		}
	}
	return moves
}

func (c *Checkers) hasCaptures(color int) bool {
	for row := 0; row < 8; row++ {
		for col := 0; col < 8; col++ {
			if c.board[row][col] == color {
				for _, m := range c.getMoves(row, col) {
					if m[2][0] != -1 {
						return true
					}
				}
			}
		}
	}
	return false
}

func (c *Checkers) makeMove(from [2]int, to [2]int, capture [2]int) {
	c.board[to[0]][to[1]] = c.board[from[0]][from[1]]
	c.board[from[0]][from[1]] = EMPTY
	if capture[0] != -1 {
		c.board[capture[0]][capture[1]] = EMPTY
	}
}

func (c *Checkers) isGameOver() bool {
	whiteMoves := c.getAllMoves(WHITE)
	blackMoves := c.getAllMoves(BLACK)
	if len(whiteMoves) == 0 || len(blackMoves) == 0 {
		return true
	}
	hasWhite := false
	hasBlack := false
	for row := 0; row < 8; row++ {
		for col := 0; col < 8; col++ {
			if c.board[row][col] == WHITE {
				hasWhite = true
			} else if c.board[row][col] == BLACK {
				hasBlack = true
			}
		}
	}
	return !hasWhite || !hasBlack
}

func (c *Checkers) getWinner() int {
	whiteMoves := c.getAllMoves(WHITE)
	blackMoves := c.getAllMoves(BLACK)
	hasWhite := false
	hasBlack := false
	for row := 0; row < 8; row++ {
		for col := 0; col < 8; col++ {
			if c.board[row][col] == WHITE {
				hasWhite = true
			} else if c.board[row][col] == BLACK {
				hasBlack = true
			}
		}
	}
	if !hasWhite || len(whiteMoves) == 0 {
		return BLACK
	}
	if !hasBlack || len(blackMoves) == 0 {
		return WHITE
	}
	return 0
}

func (c *Checkers) parseMove(s string) ([2]int, [2]int, bool) {
	parts := strings.Fields(s)
	if len(parts) != 2 {
		return [2]int{}, [2]int{}, false
	}
	if len(parts[0]) < 2 || len(parts[1]) < 2 {
		return [2]int{}, [2]int{}, false
	}
	col1 := int(parts[0][0] - 'a')
	row1 := int(parts[0][1] - '1')
	col2 := int(parts[1][0] - 'a')
	row2 := int(parts[1][1] - '1')
	if row1 < 0 || row1 > 7 || col1 < 0 || col1 > 7 || row2 < 0 || row2 > 7 || col2 < 0 || col2 > 7 {
		return [2]int{}, [2]int{}, false
	}
	return [2]int{row1, col1}, [2]int{row2, col2}, true
}

func (c *Checkers) Play() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("Добро пожаловать в мини-шашки!")
	fmt.Println("Белые (○) ходят первыми.")
	fmt.Println("Вводите ход в формате: a2 b3")
	for !c.isGameOver() {
		c.PrintBoard()
		color := c.turn
		colorName := "белых (○)"
		if color == BLACK {
			colorName = "чёрных (●)"
		}
		fmt.Printf("Ход %s.\n", colorName)
		allMoves := c.getAllMoves(color)
		if len(allMoves) == 0 {
			fmt.Println("Нет доступных ходов, игра окончена.")
			break
		}
		hasCap := c.hasCaptures(color)
		moves := allMoves
		if hasCap {
			fmt.Println("Обязательный захват!")
			moves = nil
			for _, m := range allMoves {
				if m[2][0] != -1 {
					moves = append(moves, m)
				}
			}
		}
		for {
			fmt.Print("> ")
			cmd, _ := reader.ReadString('\n')
			cmd = strings.TrimSpace(cmd)
			if cmd == "quit" {
				return
			}
			from, to, ok := c.parseMove(cmd)
			if !ok {
				fmt.Println("Неверный формат. Используйте: a2 b3")
				continue
			}
			found := -1
			for i, m := range moves {
				if m[0] == from && m[1] == to {
					found = i
					break
				}
			}
			if found == -1 {
				fmt.Println("Неверный ход. Попробуйте снова.")
				continue
			}
			c.makeMove(moves[found][0], moves[found][1], moves[found][2])
			c.turn = BLACK
			if color == BLACK {
				c.turn = WHITE
			}
			break
		}
	}
	c.PrintBoard()
	winner := c.getWinner()
	if winner == WHITE {
		fmt.Println("Победили белые (○)!")
	} else if winner == BLACK {
		fmt.Println("Победили чёрные (●)!")
	} else {
		fmt.Println("Ничья?")
	}
}

func main() {
	game := NewCheckers()
	game.Play()
}

File: checkers.rs
// checkers.rs
use std::io::{self, Write};

const EMPTY: i8 = 0;
const WHITE: i8 = 1;
const BLACK: i8 = 2;

struct Checkers {
    board: [[i8; 8]; 8],
    turn: i8,
}

impl Checkers {
    fn new() -> Self {
        let mut board = [[EMPTY; 8]; 8];
        for row in 0..8 {
            for col in 0..8 {
                if (row + col) % 2 == 1 {
                    if row < 3 {
                        board[row][col] = WHITE;
                    } else if row > 4 {
                        board[row][col] = BLACK;
                    }
                }
            }
        }
        Checkers { board, turn: WHITE }
    }

    fn print_board(&self) {
        println!("  a b c d e f g h");
        for row in 0..8 {
            print!("{} ", row + 1);
            for col in 0..8 {
                let ch = match self.board[row][col] {
                    EMPTY => '·',
                    WHITE => '○',
                    BLACK => '●',
                    _ => '?',
                };
                print!("{} ", ch);
            }
            println!();
        }
    }

    fn is_valid_pos(row: i32, col: i32) -> bool {
        row >= 0 && row < 8 && col >= 0 && col < 8
    }

    fn get_moves(&self, row: usize, col: usize) -> Vec<((usize, usize), (usize, usize), Option<(usize, usize)>)> {
        let piece = self.board[row][col];
        if piece == EMPTY {
            return vec![];
        }
        let dir: i32 = if piece == WHITE { 1 } else { -1 };
        let mut moves = Vec::new();
        let row_i = row as i32;
        let col_i = col as i32;
        // простые ходы
        for dc in &[-1, 1] {
            let nr = row_i + dir;
            let nc = col_i + dc;
            if Self::is_valid_pos(nr, nc) {
                let nr_u = nr as usize;
                let nc_u = nc as usize;
                if self.board[nr_u][nc_u] == EMPTY {
                    moves.push(((row, col), (nr_u, nc_u), None));
                }
            }
        }
        // захваты
        for dc in &[-1, 1] {
            let nr = row_i + dir * 2;
            let nc = col_i + dc * 2;
            if Self::is_valid_pos(nr, nc) {
                let mr = row_i + dir;
                let mc = col_i + dc;
                let nr_u = nr as usize;
                let nc_u = nc as usize;
                let mr_u = mr as usize;
                let mc_u = mc as usize;
                if self.board[mr_u][mc_u] != EMPTY && self.board[mr_u][mc_u] != piece && self.board[nr_u][nc_u] == EMPTY {
                    moves.push(((row, col), (nr_u, nc_u), Some((mr_u, mc_u))));
                }
            }
        }
        moves
    }

    fn get_all_moves(&self, color: i8) -> Vec<((usize, usize), (usize, usize), Option<(usize, usize)>)> {
        let mut all = Vec::new();
        for row in 0..8 {
            for col in 0..8 {
                if self.board[row][col] == color {
                    all.extend(self.get_moves(row, col));
                }
            }
        }
        all
    }

    fn has_captures(&self, color: i8) -> bool {
        for row in 0..8 {
            for col in 0..8 {
                if self.board[row][col] == color {
                    for m in self.get_moves(row, col) {
                        if m.2.is_some() {
                            return true;
                        }
                    }
                }
            }
        }
        false
    }

    fn make_move(&mut self, from: (usize, usize), to: (usize, usize), capture: Option<(usize, usize)>) {
        let (fr, fc) = from;
        let (tr, tc) = to;
        self.board[tr][tc] = self.board[fr][fc];
        self.board[fr][fc] = EMPTY;
        if let Some((cr, cc)) = capture {
            self.board[cr][cc] = EMPTY;
        }
    }

    fn is_game_over(&self) -> bool {
        let white_moves = self.get_all_moves(WHITE);
        let black_moves = self.get_all_moves(BLACK);
        if white_moves.is_empty() || black_moves.is_empty() {
            return true;
        }
        let mut has_white = false;
        let mut has_black = false;
        for row in 0..8 {
            for col in 0..8 {
                match self.board[row][col] {
                    WHITE => has_white = true,
                    BLACK => has_black = true,
                    _ => {}
                }
            }
        }
        !has_white || !has_black
    }

    fn get_winner(&self) -> i8 {
        let white_moves = self.get_all_moves(WHITE);
        let black_moves = self.get_all_moves(BLACK);
        let mut has_white = false;
        let mut has_black = false;
        for row in 0..8 {
            for col in 0..8 {
                match self.board[row][col] {
                    WHITE => has_white = true,
                    BLACK => has_black = true,
                    _ => {}
                }
            }
        }
        if !has_white || white_moves.is_empty() {
            return BLACK;
        }
        if !has_black || black_moves.is_empty() {
            return WHITE;
        }
        0
    }

    fn parse_move(s: &str) -> Option<((usize, usize), (usize, usize))> {
        let parts: Vec<&str> = s.split_whitespace().collect();
        if parts.len() != 2 {
            return None;
        }
        if parts[0].len() < 2 || parts[1].len() < 2 {
            return None;
        }
        let c1 = parts[0].chars().nth(0).unwrap();
        let r1 = parts[0].chars().nth(1).unwrap();
        let c2 = parts[1].chars().nth(0).unwrap();
        let r2 = parts[1].chars().nth(1).unwrap();
        let col1 = (c1 as u8 - b'a') as i32;
        let row1 = (r1 as u8 - b'1') as i32;
        let col2 = (c2 as u8 - b'a') as i32;
        let row2 = (r2 as u8 - b'1') as i32;
        if row1 < 0 || row1 > 7 || col1 < 0 || col1 > 7 || row2 < 0 || row2 > 7 || col2 < 0 || col2 > 7 {
            return None;
        }
        Some(((row1 as usize, col1 as usize), (row2 as usize, col2 as usize)))
    }

    fn play(&mut self) {
        println!("Добро пожаловать в мини-шашки!");
        println!("Белые (○) ходят первыми.");
        println!("Вводите ход в формате: a2 b3");
        let stdin = io::stdin();
        let mut input = String::new();
        while !self.is_game_over() {
            self.print_board();
            let color = self.turn;
            let color_name = if color == WHITE { "белых (○)" } else { "чёрных (●)" };
            println!("Ход {}.", color_name);
            let all_moves = self.get_all_moves(color);
            if all_moves.is_empty() {
                println!("Нет доступных ходов, игра окончена.");
                break;
            }
            let has_cap = self.has_captures(color);
            let moves = if has_cap {
                println!("Обязательный захват!");
                all_moves.into_iter().filter(|m| m.2.is_some()).collect()
            } else {
                all_moves
            };
            loop {
                print!("> ");
                io::stdout().flush().unwrap();
                input.clear();
                stdin.read_line(&mut input).unwrap();
                let cmd = input.trim();
                if cmd == "quit" {
                    return;
                }
                if let Some((from, to)) = Self::parse_move(cmd) {
                    let found = moves.iter().position(|m| m.0 == from && m.1 == to);
                    if let Some(idx) = found {
                        let m = &moves[idx];
                        self.make_move(m.0, m.1, m.2);
                        self.turn = if color == WHITE { BLACK } else { WHITE };
                        break;
                    } else {
                        println!("Неверный ход. Попробуйте снова.");
                    }
                } else {
                    println!("Неверный формат. Используйте: a2 b3");
                }
            }
        }
        self.print_board();
        let winner = self.get_winner();
        if winner == WHITE {
            println!("Победили белые (○)!");
        } else if winner == BLACK {
            println!("Победили чёрные (●)!");
        } else {
            println!("Ничья?");
        }
    }
}

fn main() {
    let mut game = Checkers::new();
    game.play();
}
